Fix compatibility with PPC64LE platform

etehtsea opened the pull request:
Updated to libffi master instead of latest stable due to lack of `--disable-docs` in the latter.

etehtsea commented:
Also checked that `ffi_prep_cif_var` works under windows. All specs are green.

```ruby
 RbConfig::CONFIG['host_os']
=> "mingw32"
```

tduehr commented:
Is there anything else related to `HAVE_FFI_PREP_CIF_VAR` that needs to be done?

etehtsea commented:
@tduehr  I would prefer to do checking of `ffi_prep_cif_var` existence dynamically, but I don't know any way how to do this.

etehtsea commented:
Another thing I spotted is msvc build will [no longer work as it is](https://github.com/libffi/libffi/blob/master/README#L151-L169). What do you think about dropping its support?

etehtsea commented:
I removed my commit which updates libffi and rebased to your `libffi` branch.
What else should be done to get this merged? If there is a list of supported platforms I could try to prepare compatibility matrix.

etehtsea commented:
@larskanis I've found the solution [here](https://github.com/libffi/libffi/issues/210). So this could be libffi specific issue.

etehtsea commented:
Regarding to failures on osx. It looks like this is travis-ci specific issue, because I couldn't reproduce it locally on macOS Sierra neither on ElCapitan (which used on travisci) in vbox. I think it's safe to move these builds to allowed to fail section.

```
Darwin This-MacBook-Pro.local 15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 23 18:25:34 PDT 2016; root:xnu-3248.60.10~1/RELEASE_X86_64 x86_64

This-MacBook-Pro:ffi vagrant$ rvm use 2.1
rubUsing /Users/vagrant/.rvm/gems/ruby-2.1.10
This-MacBook-Pro:ffi vagrant$ ruby -v
ruby 2.1.10p492 (2016-04-01 revision 54464) [x86_64-darwin15.0]
This-MacBook-Pro:ffi vagrant$ gem -v
2.6.12
This-MacBook-Pro:ffi vagrant$ gem install bundler
Fetching: bundler-1.15.1.gem (100%)
Successfully installed bundler-1.15.1
Parsing documentation for bundler-1.15.1
Installing ri documentation for bundler-1.15.1
Done installing documentation for bundler after 6 seconds
1 gem installed
This-MacBook-Pro:ffi vagrant$ bundle exec rake test
Could not find rake-10.5.0 in any of the sources
Run `bundle install` to install missing gems.
This-MacBook-Pro:ffi vagrant$ bundle
Fetching gem metadata from https://rubygems.org/..........
Fetching version metadata from https://rubygems.org/.
Fetching rake 10.5.0
Installing rake 10.5.0
Fetching rake-compiler-dock 0.6.1
Installing rake-compiler-dock 0.6.1
Fetching rspec-support 3.6.0
Installing rspec-support 3.6.0
Fetching diff-lcs 1.3
Installing diff-lcs 1.3
Fetching rubygems-tasks 0.2.4
Installing rubygems-tasks 0.2.4
Fetching yard 0.8.7.6
Installing yard 0.8.7.6
Using bundler 1.15.1
Fetching rake-compiler 1.0.4
Installing rake-compiler 1.0.4
Fetching rspec-core 3.6.0
Installing rspec-core 3.6.0
Fetching rspec-expectations 3.6.0
Installing rspec-expectations 3.6.0
Fetching rspec-mocks 3.6.0
Installing rspec-mocks 3.6.0
Fetching rspec 3.6.0
Installing rspec 3.6.0
Bundle complete! 7 Gemfile dependencies, 12 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
This-MacBook-Pro:ffi vagrant$ bundle exec rake test
mkdir -p build/x86_64-darwin15.0/ffi_c/2.1.10
cd build/x86_64-darwin15.0/ffi_c/2.1.10
/Users/vagrant/.rvm/rubies/ruby-2.1.10/bin/ruby -I. ../../../../ext/ffi_c/extconf.rb
checking for shlwapi.h... no
checking for rb_thread_blocking_region()... yes
checking for rb_thread_call_with_gvl()... yes
checking for rb_thread_call_without_gvl()... yes
creating extconf.h
creating Makefile
cd -
cd build/x86_64-darwin15.0/ffi_c/2.1.10
/usr/bin/make
Running autoreconf for libffi
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: running: glibtoolize --copy --force
autoreconf: running: /usr/local/Cellar/autoconf/2.69/bin/autoconf --force
autoreconf: running: /usr/local/Cellar/autoconf/2.69/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:31: installing './compile'
configure.ac:19: installing './missing'
Makefile.am: installing './depcomp'
autoreconf: Leaving directory `.'
Configuring libffi
clang: error: unsupported option '-print-multi-os-directory'
clang: error: no input files
cd "/private/tmp/ffi/build/x86_64-darwin15.0/ffi_c/2.1.10/libffi-x86_64-darwin15.0" && /Library/Developer/CommandLineTools/usr/bin/make
/Library/Developer/CommandLineTools/usr/bin/make 'AR_FLAGS=' 'CC_FOR_BUILD=' 'CFLAGS=-Wall -fexceptions' 'CXXFLAGS=-g -O2' 'CFLAGS_FOR_BUILD=' 'CFLAGS_FOR_TARGET=' 'INSTALL=/usr/local/bin/ginstall -c' 'INSTALL_DATA=/usr/local/bin/ginstall -c -m 644' 'INSTALL_PROGRAM=/usr/local/bin/ginstall -c' 'INSTALL_SCRIPT=/usr/local/bin/ginstall -c' 'JC1FLAGS=' 'LDFLAGS=' 'LIBCFLAGS=' 'LIBCFLAGS_FOR_TARGET=' 'MAKE=/Library/Developer/CommandLineTools/usr/bin/make' 'MAKEINFO=/bin/sh /private/tmp/ffi/ext/ffi_c/libffi/missing makeinfo ' 'PICFLAG=' 'PICFLAG_FOR_TARGET=' 'RUNTESTFLAGS=' 'SHELL=/bin/sh' 'exec_prefix=/usr/local' 'infodir=/usr/local/share/info' 'libdir=/usr/local/lib' 'mandir=/usr/local/share/man' 'prefix=/usr/local' 'AR=ar' 'AS=as' 'CC=gcc' 'CXX=g++' 'LD=ld' 'NM=/usr/bin/nm -B' 'RANLIB=ranlib' 'DESTDIR=' all-recursive
Making all in include
make[3]: Nothing to be done for `all'.
Making all in testsuite
make[3]: Nothing to be done for `all'.
Making all in man
make[3]: Nothing to be done for `all'.
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/prep_cif.lo /private/tmp/ffi/ext/ffi_c/libffi/src/prep_cif.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/prep_cif.c  -fno-common -DPIC -o src/.libs/prep_cif.o
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/types.lo /private/tmp/ffi/ext/ffi_c/libffi/src/types.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/types.c  -fno-common -DPIC -o src/.libs/types.o
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/raw_api.lo /private/tmp/ffi/ext/ffi_c/libffi/src/raw_api.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/raw_api.c  -fno-common -DPIC -o src/.libs/raw_api.o
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/java_raw_api.lo /private/tmp/ffi/ext/ffi_c/libffi/src/java_raw_api.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/java_raw_api.c  -fno-common -DPIC -o src/.libs/java_raw_api.o
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/closures.lo /private/tmp/ffi/ext/ffi_c/libffi/src/closures.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/closures.c  -fno-common -DPIC -o src/.libs/closures.o
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/x86/ffi64.lo /private/tmp/ffi/ext/ffi_c/libffi/src/x86/ffi64.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/x86/ffi64.c  -fno-common -DPIC -o src/x86/.libs/ffi64.o
/bin/sh ./libtool    --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src  -c -o src/x86/unix64.lo /private/tmp/ffi/ext/ffi_c/libffi/src/x86/unix64.S
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -c /private/tmp/ffi/ext/ffi_c/libffi/src/x86/unix64.S  -fno-common -DPIC -o src/x86/.libs/unix64.o
/bin/sh ./libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src   -Wall -fexceptions -c -o src/x86/ffiw64.lo /private/tmp/ffi/ext/ffi_c/libffi/src/x86/ffiw64.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -Wall -fexceptions -c /private/tmp/ffi/ext/ffi_c/libffi/src/x86/ffiw64.c  -fno-common -DPIC -o src/x86/.libs/ffiw64.o
/bin/sh ./libtool    --mode=compile gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src  -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src  -c -o src/x86/win64.lo /private/tmp/ffi/ext/ffi_c/libffi/src/x86/win64.S
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/private/tmp/ffi/ext/ffi_c/libffi -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -I. -I/private/tmp/ffi/ext/ffi_c/libffi/include -Iinclude -I/private/tmp/ffi/ext/ffi_c/libffi/src -c /private/tmp/ffi/ext/ffi_c/libffi/src/x86/win64.S  -fno-common -DPIC -o src/x86/.libs/win64.o
/bin/sh ./libtool  --tag=CC   --mode=link gcc  -Wall -fexceptions   -o libffi_convenience.la  src/prep_cif.lo src/types.lo src/raw_api.lo src/java_raw_api.lo src/closures.lo   src/x86/ffi64.lo src/x86/unix64.lo src/x86/ffiw64.lo src/x86/win64.lo
libtool: link: ar cru .libs/libffi_convenience.a src/.libs/prep_cif.o src/.libs/types.o src/.libs/raw_api.o src/.libs/java_raw_api.o src/.libs/closures.o src/x86/.libs/ffi64.o src/x86/.libs/unix64.o src/x86/.libs/ffiw64.o src/x86/.libs/win64.o
libtool: link: ranlib .libs/libffi_convenience.a
libtool: link: ( cd ".libs" && rm -f "libffi_convenience.la" && ln -s "../libffi_convenience.la" "libffi_convenience.la" )
/bin/sh ./libtool  --tag=CC   --mode=link gcc  -Wall -fexceptions -no-undefined -version-info `grep -v '^#' /private/tmp/ffi/ext/ffi_c/libffi/libtool-version`     -o libffi.la -rpath /usr/local/lib src/prep_cif.lo src/types.lo src/raw_api.lo src/java_raw_api.lo src/closures.lo  src/x86/ffi64.lo src/x86/unix64.lo src/x86/ffiw64.lo src/x86/win64.lo
libtool: link: gcc -dynamiclib  -o .libs/libffi.7.dylib  src/.libs/prep_cif.o src/.libs/types.o src/.libs/raw_api.o src/.libs/java_raw_api.o src/.libs/closures.o src/x86/.libs/ffi64.o src/x86/.libs/unix64.o src/x86/.libs/ffiw64.o src/x86/.libs/win64.o      -install_name  /usr/local/lib/libffi.7.dylib -compatibility_version 9 -current_version 9.0 -Wl,-single_module
ld: warning: could not create compact unwind for _ffi_call_unix64: does not use RBP or RSP based frame
libtool: link: (cd ".libs" && rm -f "libffi.dylib" && ln -s "libffi.7.dylib" "libffi.dylib")
libtool: link: ( cd ".libs" && rm -f "libffi.la" && ln -s "../libffi.la" "libffi.la" )
compiling ../../../../ext/ffi_c/AbstractMemory.c
compiling ../../../../ext/ffi_c/ArrayType.c
compiling ../../../../ext/ffi_c/Buffer.c
compiling ../../../../ext/ffi_c/Call.c
../../../../ext/ffi_c/Call.c:355:5: warning: implicit declaration of function 'rb_thread_call_without_gvl' is invalid in C99 [-Wimplicit-function-declaration]
    rbffi_thread_blocking_region(call_blocking_function, data, (void *) -1, NULL);
    ^
../../../../ext/ffi_c/Thread.h:78:39: note: expanded from macro 'rbffi_thread_blocking_region'
# define rbffi_thread_blocking_region rb_thread_call_without_gvl
                                      ^
1 warning generated.
compiling ../../../../ext/ffi_c/ClosurePool.c
compiling ../../../../ext/ffi_c/DataConverter.c
compiling ../../../../ext/ffi_c/DynamicLibrary.c
compiling ../../../../ext/ffi_c/ffi.c
compiling ../../../../ext/ffi_c/Function.c
../../../../ext/ffi_c/Function.c:563:9: warning: implicit declaration of function 'rb_thread_call_without_gvl' is invalid in C99 [-Wimplicit-function-declaration]
        rb_thread_call_without_gvl(async_cb_wait, &w, async_cb_stop, &w);
        ^
../../../../ext/ffi_c/Function.c:948:17: warning: 'ffi_prep_closure' is deprecated: use ffi_prep_closure_loc instead [-Wdeprecated-declarations]
    ffiStatus = ffi_prep_closure(code, &fnInfo->ffi_cif, callback_invoke, closure);
                ^
/private/tmp/ffi/build/x86_64-darwin15.0/ffi_c/2.1.10/libffi-x86_64-darwin15.0/include/ffi.h:314:1: note: 'ffi_prep_closure' has been explicitly marked deprecated here
ffi_prep_closure (ffi_closure*,
^
2 warnings generated.
compiling ../../../../ext/ffi_c/FunctionInfo.c
compiling ../../../../ext/ffi_c/LastError.c
compiling ../../../../ext/ffi_c/LongDouble.c
compiling ../../../../ext/ffi_c/MappedType.c
compiling ../../../../ext/ffi_c/MemoryPointer.c
compiling ../../../../ext/ffi_c/MethodHandle.c
compiling ../../../../ext/ffi_c/Platform.c
compiling ../../../../ext/ffi_c/Pointer.c
compiling ../../../../ext/ffi_c/Struct.c
compiling ../../../../ext/ffi_c/StructByReference.c
compiling ../../../../ext/ffi_c/StructByValue.c
compiling ../../../../ext/ffi_c/StructLayout.c
compiling ../../../../ext/ffi_c/Thread.c
compiling ../../../../ext/ffi_c/Type.c
compiling ../../../../ext/ffi_c/Types.c
compiling ../../../../ext/ffi_c/Variadic.c
linking shared-object ffi_c.bundle
ld: warning: could not create compact unwind for _ffi_call_unix64: does not use RBP or RSP based frame
cd -
<..>
/Users/vagrant/.rvm/rubies/ruby-2.1.10/bin/ruby -I/Users/vagrant/.rvm/gems/ruby-2.1.10/gems/rspec-core-3.6.0/lib:/Users/vagrant/.rvm/gems/ruby-2.1.10/gems/rspec-support-3.6.0/lib /Users/vagrant/.rvm/gems/ruby-2.1.10/gems/rspec-core-3.6.0/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb --color -w --format documentation
Run options: exclude {:broken=>true}
<..>
Finished in 2.32 seconds (files took 1.12 seconds to load)
4845 examples, 0 failures
```

etehtsea commented:
OK, I'm done with with this PR.

tduehr commented:
this problem is due to not checking out the submodule before building libffi. The tests need to remain and be fixed.

tduehr commented:
With libffi as a submodule, it must be updated before compiling.

`git submodule update`

etehtsea commented:
@tduehr I'm not sure that I understood you. According to [travis log](https://travis-ci.org/etehtsea/ffi/jobs/243185975) it automatically fetch submodules. What else should be done?

etehtsea commented:
@tduehr I fixed building on 2.1/2.2 on macOS

etehtsea commented on Rakefile:
```diff
@@ -191,7 +191,7 @@ if USE_RAKE_COMPILER
   task "gem:windows" do
     require "rake_compiler_dock"
-    RakeCompilerDock.sh "bundle && rake cross native gem MAKE='nice make -j`nproc`'"
+    RakeCompilerDock.sh "sudo apt-get update && sudo apt-get install -y libltdl-dev && bundle && rake cross native gem MAKE='nice make -j`nproc`'"
```
https://github.com/ffi/ffi/pull/577#issuecomment-308498546

I'm not sure about it.
Anyway, this could be addressed as part of another issue if needed.

tduehr replied:
With the stuff you found in libffi, I'm comfortable leaving it as is for now. Thanks for the work on this.
